packages/dds/tree/src/feature-libraries/defaultFieldKinds.ts

```diff
@@ -189,18 +194,148 @@ export const counter: FieldKind = new FieldKind(
+        if (changes.length === 0) {
+            return {};
+        }
+        let newValue: JsonableTree | undefined;
+        const childChanges: NodeChangeset[] = [];
+        for (const change of changes) {
+            if (change.value !== undefined) {
+                newValue = change.value;
+
+                // The previous changes applied to a different value, so we discard them.
+                // TODO: Consider if we should represent muted changes
+                childChanges.length = 0;
+            }
+
+            if (change.changes !== undefined) {
+                childChanges.push(change.changes);
+            }
+        }
+
+        const composed: ValueChangeset = {};
+        if (newValue !== undefined) {
+            composed.value = newValue;
+        }
+
+        if (childChanges.length > 0) {
+            composed.changes = composeChildren(childChanges);
+        }
+
+        return composed;
+    },
+
+    invert: (change: ValueChangeset, invertChild: NodeChangeInverter): ValueChangeset => {
+        // TODO: Handle the value inverse
+        const inverse: ValueChangeset = {};
+        if (change.changes !== undefined) {
+            inverse.changes = invertChild(change.changes);
+        }
+        return inverse;
+    },
+
+    rebase: (change: ValueChangeset, over: ValueChangeset, rebaseChild: NodeChangeRebaser): ValueChangeset => {
+        if (change.changes === undefined || over.changes === undefined) {
+            return change;
+        }
+        return { ...change, changes: rebaseChild(change.changes, over.changes) };
+    },
+};
+
+interface EncodedValueChangeset {
+    value?: JsonableTree;
+    changes?: JsonCompatibleReadOnly;
+}
+
+const valueFieldEncoder: FieldChangeEncoder<ValueChangeset> = {
+    encodeForJson: (formatVersion: number, change: ValueChangeset, encodeChild: NodeChangeEncoder) => {
+        const encoded: EncodedValueChangeset & JsonCompatibleReadOnly = {};
+        if (change.value !== undefined) {
+            encoded.value = change.value;
+        }
+
+        if (change.changes !== undefined) {
+            encoded.changes = encodeChild(change.changes);
+        }
+
+        return encoded;
+    },
+
+    decodeJson: (formatVersion: number, change: JsonCompatibleReadOnly, decodeChild: NodeChangeDecoder) => {
+        const encoded = change as EncodedValueChangeset;
+        const decoded: ValueChangeset = {};
+        if (encoded.value !== undefined) {
+            decoded.value = encoded.value;
+        }
+
+        if (encoded.changes !== undefined) {
+            decoded.changes = decodeChild(encoded.changes);
+        }
+
+        return decoded;
+    },
+};
+
+export interface ValueFieldEditor extends FieldEditor<ValueChangeset> {
+    /**
+     * Creates a change which replaces the current value of the field with `newValue`.
+     * `newValue` should not be mutated once passed to this call.
+     */
+    set(newValue: JsonableTree): ValueChangeset;
```

Comment:
Why `JsonableTree` and not `ITreeCursor` (which is what `SequenceEditBuilder.insert` takes)?